docs: update observability/telemetry overview pages (#6563)
apollographql/router: docs/source/routing/observability/telemetry.mdx

    end

    apms["APM, agent,<br/>or collector"]

>     exporters--"OLTP"-->apms

Reviewer: OpenTeLemetryProtocol (OTLP): https://opentelemetry.io/docs/specs/otel/protocol/

- Prometheus
- OpenTelemetry Collector
- Datadog
- New Relic
- Jaeger
- Zipkin

### Attributes and selectors

Attributes and selectors are key-value pairs that add contextual information about the router request lifecycle to telemetry data. You can annotate events, metrics, and spans with them and then use them to filter and group data in your APMs.

The router supports a set of standard attributes from [OpenTelemetry semantic conventions](https://opentelemetry.io/docs/specs/semconv/). Example attributes include:

- HTTP status code  
- GraphQL operation name  
- Subgraph name

Selectors allow you define custom data points based on the router's request lifecycle. 

|  | Description |
| ----- | ----- |
| **Attribute** | Standard data points that can be attached to spans, instruments, and events. |
| **Selector** | Custom data points extracted from the router's request lifecycle, tailored to specific needs. |

## Router telemetry signals

The router supports three signal types for collecting and exporting telemetry:

<table>
  <thead>
    <tr>
      <th>Signal</th>
      <th>Description</th>
    </tr>
  </thead>
  <tbody>
    <tr>

>       <td><a href="/graphos/routing/observability/telemetry#logs-and-events">Logs and events</a></td>

Author: Will remove all links in a separate commit

>       <td><a href="/graphos/routing/observability/telemetry#logs-and-events">Logs and events</a></td>

Author: https://github.com/apollographql/router/pull/6563/commits/7b7206f98eafa43084de6f3a10d74cabffa876f8

```mermaid
flowchart LR
    Router --"Emits logs in<br/>text or JSON format"--> stdout
    stdout --"Exports logs"--> log_store
    log_store[("Log store")]
```

#### Example log configuration

This configuration snippet enables stdout logging in JSON:

```yaml title="router.yaml"
telemetry:
  exporters:
    logging:
      stdout:
        enabled: true
        format: json
```

### Metrics and instruments

Metrics are measurements of the router's behavior that are collected and often analyzed over time to identify trends. Examples of router metrics include the number of incoming HTTP requests and the time spent processing a request. 

Instruments define _how_ to collect and report metrics. Different kinds of instruments include counters, gauges, and histograms. Given the metric “number of incoming HTTP requests”, different instruments collect a count, histogram, and gauge of this metric.

> 

Reviewer 2: One small suggestion to augment this a bit:

> different instruments collect a count, histogram, and gauge of this metric

> For example, given the metric 'number of incoming HTTP requests,' a counter records the total number of requests, a histogram captures the distribution of request counts over time, and a gauge provides a snapshot of the current request count at a given moment."

Author: https://github.com/apollographql/router/pull/6563/commits/cc6eb4c93dec4574b4139ec00c5fa6f408a5bb57

#### Instrument types

Metric instruments fall into three categories:

<table>
  <thead>
    <tr>
      <th>Instrument Type</th>
      <th>Description</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><strong>Router instruments</strong></td>
      <td>
        Standard instruments for the router's request life cycle, including:
        <ul>
          <li>The number of HTTP requests by HTTP status</li>
          <li>A histogram of HTTP router request duration</li>
        </ul>
      </td>
    </tr>
    <tr>
      <td><strong>OTEL instruments</strong></td>
      <td>
        Standard OpenTelemetry instruments around the HTTP lifecycle, including:
        <ul>
          <li>The number of active requests in flight</li>
          <li>A histogram of request body sizes</li>
        </ul>
      </td>
    </tr>
    <tr>
      <td><strong>Custom instrument</strong></td>
      <td>
        Custom instruments defined in the router's request life cycle.
      </td>

>     </tr>

Author: https://github.com/apollographql/router/pull/6563/commits/cc6eb4c93dec4574b4139ec00c5fa6f408a5bb57

  </tbody>
</table>

#### Example instrument configuration

This configuration snippet enables OTEL instrumentation for a histogram of request body sizes:

```yaml title="router.yaml"
telemetry:
  instrumentation:
    instruments:
      router:
        http.server.request.body.size: true
```

See [Instruments](/router/configuration/telemetry/instrumentation/instruments) for an overview of available instruments and a guide for configuring and customizing instruments.

#### Metric exporters

> The router sends [operation](/graphos/platform/insights/sending-operation-metrics#from-the-apollo-router-or-apollo-server) and [field usage metrics](/graphos/platform/insights/sending-operation-metrics#from-the-apollo-router) metrics to GraphOS. Additionally, metric exporters can consume and monitor other metrics.

Reviewer 3: Need to distinguish the special case of operation metrics reported to GraphOS, as they're not one of the exporters configurable by `telemetry.exporters.*`

```suggestion
In addition to the [operation metrics](/graphos/platform/insights/sending-operation-metrics#from-the-apollo-router-or-apollo-server) and [field usage metrics](/graphos/platform/insights/sending-operation-metrics#from-the-apollo-router) that GraphOS Router sends to GraphOS, you can configure the router with metric exporters for other observability tools and APMs.
```

```mermaid
flowchart LR
    Router --"OTEL<br/>metrics"--> APM
    Router --"Usage/Performance<br/>metrics"--> GraphOS
```

> This configuration snippet enables exporting metrics to Prometheus:

Reviewer 3: Is Datadog the most popular APM now? Should update the example for whatever is the most used

Author: I used DataDog for the next example to showcase various APMs, but I'm fine with making them all DataDog if that's preferred

Reviewer 2: Datadog is definitely the most common but we've seen it all. Dynatrace, New Relic, Grafana, etc.

Author: Thanks, will keep as is unless folks object.

```yaml title="router.yaml"
telemetry:
  exporters:
     metrics:
       prometheus:
         enabled: true
         listen: 127.0.0.1:9090
         path: /metrics
```

Learn more about [sending metrics to Prometheus](/graphos/reference/router/telemetry/metrics-exporters/prometheus) and [metric exporters](/graphos/reference/router/telemetry/metrics-exporters/overview) in general.

### Traces and spans

> Traces monitor the flow of a request through the router. A trace is composed of [spans](/router/configuration/telemetry/instrumentation/spans). A span captures a request's duration as it flows through the router request lifecycle. Spans may include contextual information about the request, such as the HTTP status code or the name of the subgraph being queried.

Reviewer 3: ```suggestion
Traces help you monitor the flow of a request through the router. A trace is composed of [spans](/router/configuration/telemetry/instrumentation/spans). A span captures a request's duration as it flows through the router request lifecycle. Spans may include contextual information about the request, such as the HTTP status code or the name of the subgraph being queried.
```

```mermaid
flowchart LR
    Router --"OTEL<br/>traces"--> APM
    Router --"FTV1 Data"--> GraphOS
```

This configuration snippet enables
- setting attributes that Datadog uses to organize its APM view
- exporting traces to a Datadog agent:

```yaml title="router.yaml"
telemetry:
  instrumentation:
    spans:
      mode: spec_compliant
      router:
        attributes:
          otel.name: router
          operation.name: "router"
          resource.name:
            request_method: true
      supergraph:
        attributes:
          otel.name: supergraph
          operation.name: "supergraph"
          resource.name:
            operation_name: string
      subgraph:
        attributes:
          otel.name: subgraph
          operation.name: "subgraph"
          resource.name:
            subgraph_operation_name: string
  exporters:
    tracing:
      otlp:
        enabled: true
        endpoint: "${env.DATADOG_AGENT_HOST}:4317"
```

Learn more about [sending traces to DataDog](/graphos/reference/router/telemetry/trace-exporters/datadog) and [trace exporters](/graphos/reference/router/telemetry/trace-exporters/overview) in general.

## Best practices

> ### Collect exactly the telemetry you need

Reviewer 3: ```suggestion
### Collecting exactly the telemetry you need
```

Effective telemetry provides just the right amount and granularity of information to maintain your graph. Too much data can overwhelm your system, for example, with high cardinality metrics. Too little may not provide enough information to debug issues.

Specific events that need to be captured—and the conditions under which they need to be captured—can change as client applications and graphs change. Different environments, such as production and development, can have different observability requirements.

Router telemetry is customizable to meet the observability needs of different graphs. Keep in mind your particular environments' and graphs' requirements when configuring your telemetry.

> #### Conditions

Reviewer 3: ```suggestion
#### Setting conditions for collecting telemetry
```

> You can set conditions for instruments and events to only collect telemetry data when necessary. This configuration snippet enables only collecting the configured telemetry data when the `request_header` is equal to "example-value":

Reviewer 3: ```suggestion
You can set [conditions](/graphos/reference/router/telemetry/instrumentation/conditions) for instruments and events to only collect telemetry data when necessary. This configuration snippet enables only collecting the configured telemetry data when the `request_header` is equal to "example-value":
```

You can use [standard attributes](/router/configuration/telemetry/instrumentation/standard-attributes) or [selectors](/router/configuration/telemetry/instrumentation/selectors) as span attributes.
Specific events that need to be captured—and the conditions under which they need to be captured—can change as client applications and graphs change. Different environments, such as production and development, can have different observability requirements.

<PremiumFeature>
Router telemetry is customizable to meet the observability needs of different graphs. Keep in mind your particular environments' and graphs' requirements when configuring your telemetry.

[Custom attributes for spans](/router/configuration/telemetry/instrumentation/spans/#attributes) require a GraphOS [Dedicated or Enterprise plan](https://www.apollographql.com/pricing#observability).
#### Conditions

</PremiumFeature>
You can set conditions for instruments and events to only collect telemetry data when necessary. This configuration snippet enables only collecting the configured telemetry data when the `request_header` is equal to "example-value":

> ## Best practices

Author: Comment from @devfreddy-apollo on this section:

> Yes! This is an area we should expand on for sure. The router team has opinions, CS should have opinions as well. I have some I've started to aggregate but we should crowd source this too. IMO this is where customers get value from all of these great O11y features, and if we don't tell them or do it by default they aren't going to do it.

I agree that we should add more best practices and recipes but would like to get this out-the-door for now. As we expand about best practices, we can move this subsection to its own page, and perhaps eventually to its own set of pages.

> ### Balancing telemetry and router performance

Author: Comment from @devfreddy-apollo on this section:

> The router team likely has some actual stats on this impact, but it's not something I've seen with customers at all. The router is performant enough to add A LOT of telemetry before this is an issue. It's good to document, but I'm not sure it needs to be prioritized in an opening primer. That might be my personal opinion, though.

This section was originally written by the router team, so I'm inclined to keep it. @BrynCooke would you be able to comment on whether this is worth calling out upfront?
